include/range/v3/at.hpp

```diff
@@ -14,35 +15,64 @@
 #ifndef RANGES_V3_AT_HPP
 #define RANGES_V3_AT_HPP
 
+#include <stdexcept>
 #include <range/v3/range_fwd.hpp>
 #include <range/v3/begin_end.hpp>
 #include <range/v3/range_traits.hpp>
 #include <range/v3/range_concepts.hpp>
-#include <range/v3/utility/iterator.hpp>
+#include <range/v3/distance.hpp>
 #include <range/v3/utility/static_const.hpp>
+#include <range/v3/index.hpp>
+
+RANGES_DISABLE_WARNINGS
+RANGES_DIAGNOSTIC_IGNORE_SIGN_CONVERSION
 
 namespace ranges
 {
     inline namespace v3
     {
+        /// Checked indexed range access.
+        ///
         /// \ingroup group-core
         struct at_fn
         {
             /// \return `begin(rng)[n]`
             template<typename Rng,
-                CONCEPT_REQUIRES_(RandomAccessRange<Rng>())>
+                CONCEPT_REQUIRES_(RandomAccessRange<Rng>() && SizedRange<Rng>())>
             RANGES_CXX14_CONSTEXPR
-            auto operator()(Rng &&rng, range_difference_type_t<Rng> n) const ->
+            auto operator()(Rng &&rng, range_size_type_t<Rng> n) const ->
```

Comment:
Two overloads for distinct integral types will result in nasty ambiguities for other arithmetic types. What about:
```c++
template<typename Rng,
    CONCEPT_REQUIRES_(RandomAccessRange<Rng>())>
auto operator()(Rng &&rng, range_difference_type_t<Rng> n) const
RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
(
    (RANGES_EXPECT(!SizedRange<Rng>() || n < ranges::distance(rng)),
     ranges::begin(rng)[n])
)
template<typename Rng, typename T, typename Fn = at_fn,
    CONCEPT_REQUIRES_(RandomAccessRange<Rng>() &&
        ConvertibleTo<T, range_difference_type_t<Rng>>())>
auto operator()(Rng &&rng, T &&t) const
RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
(
    Fn{}((Rng &&) rng, static_cast<range_difference_type_t<Rng>>((T &&) t))
)
```